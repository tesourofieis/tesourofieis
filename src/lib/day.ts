import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  PATTERN_ADVENT,
  PATTERN_EASTER,
  PATTERN_EPIPHANY,
  PATTERN_LENT,
  PATTERN_POST_EPIPHANY_SUNDAY,
  PATTERN_PRE_LENTEN,
  TABLE_OF_PRECEDENCE,
  TEMPORA_RANK_MAP,
  TITLES,
  TYPE_TEMPORA,
} from "./constants.ts";

// """ Class used to keep `Observance` objects for particular days of Missal.
//
// It contains three lists: `tempora`, `celebration` and `commemoration`.
// On Missal's creation the lists are filled in so that `tempora` always contains `Observance` representing
// given variable day, `celebration` contains an `Observance`s to be celebrated in this day and
// `commemoration` contains zero or more `Observance`s that should be commemorated with the main celebration.
// """
class Day {
  date: string;
  tempora: Observance[] = [];
  celebration: Observance[] = [];
  commemoration: Observance[] = [];

  constructor(date: string) {
    this.date = date;
  }

  get all(): Observance[] {
    return this.tempora.concat(this.celebration).concat(this.commemoration);
  }

  getTemporaName() {
    return this.tempora[0]?.title;
  }

  serialize() {
    const serialized = {};
    const containers = ["tempora", "celebration", "commemoration"];

    for (const container of containers) {
      serialized[container] = this[container].map((item: Day) =>
        item.serialize(),
      );
    }

    return serialized;
  }
}

class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  link: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, Number.parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.link = `missal/${this.getTempora(this.name)}/${this.name}`;
    this.title = TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      if (name.replace(/^.*-(\d+).*$/, "$1") === "10-Dur") {
        this.weekday = 0;
      }
      this.weekday = Number.parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10);
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  private getTempora(name: string): string {
    if (this.flexibility === "tempora") {
      if (name.includes("nat")) {
        return "natal";
      }
      if (name.includes("adv")) {
        return "advento";
      }
      if (name.includes("pasc")) {
        return "pascoa";
      }
      if (name.includes("epi")) {
        return "epifania";
      }
      if (name.includes("quadp")) {
        return "pre-quaresma";
      }
      if (name.includes("quad")) {
        return "quaresma";
      }
      if (name.includes("pent")) {
        return "pentecostes";
      }
    }

    if (this.flexibility === "santos") {
      console.log(`Matched Santos: ${name}`);
      return "santos";
    }

    if (this.flexibility === "commune") {
      return "comum";
    }

    if (this.flexibility === "votivas") {
      return "votivas";
    }

    console.log(`Checking name: ${name}, flexibility: ${this.flexibility}`);
    throw Error(`No directory for ${name}, flexibility , ${this.flexibility}`);
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      link: this.link,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}

export { Day, Observance };
