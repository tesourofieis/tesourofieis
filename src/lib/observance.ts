import { getDate, getDay, getMonth } from "date-fns";
import {
  COMMUNE_C_10A,
  COMMUNE_C_10B,
  COMMUNE_C_10C,
  COMMUNE_C_10PASC,
  COMMUNE_C_10T,
  LANGUAGE,
  PATTERN_POST_EPIPHANY_SUNDAY,
  TABLE_OF_PRECEDENCE,
  TEMPORA_PENT23_0,
  TEMPORA_RANK_MAP,
  TYPE_TEMPORA,
  WEEKDAY_MAPPING,
  getTranslation,
} from "./constants.ts";
import { Proper, ProperConfig } from "./proper.ts";
import { ProperParser } from "./properParser.ts";

export class Observance {
  date: string;
  flexibility: string;
  name: string;
  rank: number;
  colors: string[];
  id: string;
  title: string;
  priority: number | null;
  weekday: number | null;

  // """ Build an Observance out of identifier and calendar date
  //
  // :param observance_id: observance's identifier in format
  //                <flexibility>:<identifier>:<rank>
  // :type observance_id: string
  // :param date_: specific date in which the observance is supposed
  //             to be placed. For some Sancti days its rank (class)
  //             depends on which calendar day they occur.
  // :type date_: `date ` object
  // """
  constructor(observanceId: string, date_: string) {
    const [flexibility, name, rank, color] = observanceId.split(":");
    this.date = date_;
    this.flexibility = flexibility;
    this.name = name;
    this.rank = this.calcRank(observanceId, parseInt(rank, 10));
    this.colors = Array.from(color);
    this.id = `${this.flexibility}:${this.name}:${this.rank}:${color}`;
    this.title = getTranslation(LANGUAGE).TITLES[observanceId];

    // Determine the weekday attribute based on the type of observance
    if (
      flexibility === TYPE_TEMPORA &&
      ![
        COMMUNE_C_10A,
        COMMUNE_C_10B,
        COMMUNE_C_10C,
        COMMUNE_C_10PASC,
        COMMUNE_C_10T,
      ].includes(observanceId)
    ) {
      this.weekday =
        WEEKDAY_MAPPING[parseInt(name.replace(/^.*-(\d+).*$/, "$1"), 10)];
    } else {
      this.weekday = getDay(this.date);
    }

    this.priority = this.calcPriority();
  }

  getProper(config?: ProperConfig) {
    const date = this.date;
    const proper = new ProperParser(this.id, config).parse();
    if (this.id.match(PATTERN_POST_EPIPHANY_SUNDAY) && getMonth(date) >= 9) {
      this.adjustSundayShiftedFromPostEpiphany(proper);
    }
    return proper;
  }

  hasProper(): boolean {
    return new ProperParser(this.id).properExists();
  }

  private calcRank(observanceId: string, originalRank: number): number {
    for (const caseItem of TEMPORA_RANK_MAP) {
      if (
        getMonth(this.date) === caseItem.month &&
        getDate(this.date) === caseItem.day &&
        new RegExp(caseItem.pattern).test(observanceId)
      ) {
        return caseItem.rank;
      }
    }
    return originalRank;
  }

  private calcPriority() {
    for (let priority = 0; priority < TABLE_OF_PRECEDENCE.length; priority++) {
      if (this.id.match(TABLE_OF_PRECEDENCE[priority])) {
        return priority;
      }
    }
  }

  private adjustSundayShiftedFromPostEpiphany(propers: Proper[]): void {
    const properSunday23PostPentecost = new ProperParser(
      TEMPORA_PENT23_0,
    ).parse();
    for (let i = 0; i < propers.length; i++) {
      const proper = propers[i];
      for (const section of [
        "INTROIT",
        "GRADUALE",
        "OFFERTORIUM",
        "COMMUNIO",
      ]) {
        proper.setSection(
          section,
          properSunday23PostPentecost[i][0]?.getSection(section),
        );
      }
    }
  }

  serialize() {
    return {
      id: this.id,
      rank: this.rank,
      title: this.title,
      colors: this.colors,
    };
  }
}
