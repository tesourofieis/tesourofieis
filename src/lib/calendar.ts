import { Day, Observance } from "./day.ts";
import { yyyyMMDD } from "./utils.ts";

import {
  ADVENT,
  EMBER_DAYS_SEPTEMBER,
  FROM_PRE_LENT_TO_POST_PENTECOST,
  LOCAL,
  NATIVITY_OCTAVE_SUNDAY,
  OUTRO,
  POST_EPIPHANY,
  SANCTI,
  SANCTI_10_DUr,
  TEMPORA_NAT2_0,
  WEEK_24_AFTER_PENTECOST,
} from "./constants.ts";
import { rules } from "./rules.ts";

import { UTCDate } from "@date-fns/utc";
import {
  addDays,
  format,
  getDate,
  getDay,
  getMonth,
  getYear,
  isSameDay,
  isSunday,
  nextSunday,
  nextWednesday,
  previousSaturday,
  previousSunday,
  subDays,
} from "date-fns";

class Calendar {
  private container: Map<string, Day>;
  public year: number;

  constructor(year: number) {
    this.year = year;
    this.container = new Map();
    this.buildEmptyCalendar(year);
    this.create();
  }

  private buildEmptyCalendar(year: number): void {
    let date = new UTCDate(year, 0, 1);

    while (date.getFullYear() === year) {
      this.container.set(yyyyMMDD(date), new Day(yyyyMMDD(date)));
      date = addDays(date, 1);
    }
  }

  private create() {
    this.fillInTemporaDays();
    this.fillInSanctiDays();
    this.resolveConcurrency();
  }

  private fillInTemporaDays() {
    // """
    // Days depending on variable date, such as Easter or Advent
    // """
    // # Inserting blocks
    this.insertBlock(this.calcHolyFamily(this.year), POST_EPIPHANY);
    this.insertBlock(
      this.calcSeptuagesima(this.year),
      FROM_PRE_LENT_TO_POST_PENTECOST,
    );
    this.insertBlock(
      this.calcSaturdayBefore24SundayAfterPentecost(this.year),
      POST_EPIPHANY,
      true,
      false,
      this.calcFirstAdventSunday(this.year),
    );
    this.insertBlock(
      this.calc24SundayAfterPentecost(this.year),
      WEEK_24_AFTER_PENTECOST,
    );
    this.insertBlock(
      this.calcFirstAdventSunday(this.year),
      ADVENT,
      false,
      false,
      new UTCDate(this.year, 11, 23),
    );
    this.insertBlock(
      this.calcEmberWednesdaySeptember(this.year),
      EMBER_DAYS_SEPTEMBER,
    );

    // # Inserting single days
    const holyNameDate = this.calcHolyName(this.year);

    this.container.get(yyyyMMDD(holyNameDate)).celebration = [
      new Observance(TEMPORA_NAT2_0, yyyyMMDD(holyNameDate)),
    ];

    const christKingDate = this.calcChristKing(this.year);
    if (this.container.get(yyyyMMDD(christKingDate))) {
      this.container.get(yyyyMMDD(christKingDate)).celebration = [
        new Observance(SANCTI_10_DUr, yyyyMMDD(christKingDate)),
      ];
    }

    const christmasOctaveSunday = this.calcSundayChristmasOctave(this.year);
    if (christmasOctaveSunday) {
      this.insertBlock(christmasOctaveSunday, NATIVITY_OCTAVE_SUNDAY);
    }
  }

  private insertBlock(
    date: Date,
    block: string[],
    reverse = false,
    overwrite = true,
    stopDate?: Date,
  ) {
    if (reverse) {
      // TODO: use toReversed in order not to mutate the original
      // or slice
      block = block.slice().reverse();
    }

    for (const [ii, observanceIds] of block.entries()) {
      const currentDate = addDays(new UTCDate(date), reverse ? -ii : ii);

      if (!observanceIds) {
        continue;
      }

      const dateKey = yyyyMMDD(currentDate);

      if (this.container.get(dateKey)?.celebration.length && !overwrite) {
        break;
      }

      if (isSameDay(stopDate, subDays(currentDate, 1))) {
        break;
      }

      if (this.container.get(dateKey)) {
        this.container.get(dateKey).tempora = [
          new Observance(observanceIds, dateKey),
        ];
        this.container.get(dateKey).celebration = [
          ...this.container.get(dateKey).tempora,
        ];
      }
    }
  }

  private fillInSanctiDays() {
    for (const [date] of this.container.entries()) {
      const dateId = format(date, "MM-dd");
      const days = SANCTI.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      this.container.get(date)?.celebration.push(...days);

      const local = LOCAL.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (local.length) {
        this.container.get(date)?.local.push(...local);
      }

      const outro = OUTRO.filter((ii) => ii.startsWith(`santos:${dateId}`)).map(
        (ii) => new Observance(ii, date),
      );

      if (outro.length) {
        this.container.get(date)?.outro.push(...outro);
      }
    }
  }

  // Apply `kalendar.rules.*` to the initially instantiated Calendar to fix the situations
  // where more than one Observance falls in the same day.
  private resolveConcurrency() {
    const shiftedAll: { [date: string]: Observance[] } = {};

    for (const [date] of this.container.entries()) {
      const [celebration, commemoration, shifted] = this.applyRules(
        date,
        shiftedAll[date] || [],
      );

      this.container.get(date).celebration = celebration as Observance[];
      this.container.get(date).commemoration = commemoration as Observance[];

      if (shifted) {
        // @ts-ignore
        for (const [k, v] of shifted) {
          if (!shiftedAll[k]) {
            shiftedAll[k] = [];
          }
          shiftedAll[k].push(...(Array.isArray(v) ? v : [v]));
        }
      }
    }
  }

  private applyRules(date: string, shifted: Observance[]) {
    for (const rule of rules) {
      const results = rule(
        this,
        date,
        this.container.get(date)?.tempora,
        this.container.get(date).celebration.concat(shifted),
      );

      if (results) {
        return results;
      }
    }

    return [this.container.get(date).celebration, [], []];
  }

  private calcEasterSunday(year: number): Date {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const month = Math.floor((h + l - 7 * m + 114) / 31);
    const day = ((h + l - 7 * m + 114) % 31) + 1;
    const easter = new UTCDate(year, month - 1, day);

    return easter;
  }

  private calcHolyFamily(year: number): Date {
    // Feast of the Holy Family - First Sunday after Epiphany (06 January).
    const d = new UTCDate(year, 0, 6); // Months are zero-indexed in JavaScript

    return nextSunday(d);
  }

  private calcSeptuagesima(year: number): Date {
    // """ Septuagesima Sunday.
    //
    // Beginning of the pre-Lenten season (Shrovetide).
    // It's 63 days before Ressurection, ninth Sunday before Easter, the third before Ash Wednesday.
    // First day of the Ressurection Sunday - related block.
    // """
    const easter = this.calcEasterSunday(year);
    return subDays(easter, 63);
  }

  private calcFirstAdventSunday(year: number): Date {
    // """
    // First Sunday of Advent - November 27 if it's Sunday, otherwise closest Sunday.
    // """
    const christmasDay = new UTCDate(year, 11, 25);
    const weekday = getDay(christmasDay);

    switch (weekday) {
      case 0: // Sunday
        return new UTCDate(getYear(christmasDay), 10, 27);
      case 1: // Monday
        return new UTCDate(getYear(christmasDay), 11, 3);
      case 2: // Tuesday
        return new UTCDate(getYear(christmasDay), 11, 2);
      case 3: // Wednesday
        return new UTCDate(getYear(christmasDay), 11, 1);
      case 4: // Thursday
        return new UTCDate(getYear(christmasDay), 10, 30);
      case 5: // Friday
        return new UTCDate(getYear(christmasDay), 10, 29);
      default: // Saturday
        return new UTCDate(getYear(christmasDay), 10, 28);
    }
  }

  private calc24SundayAfterPentecost(year: number): Date {
    // """ 24th Sunday after Pentecost.
    //
    // Last Sunday before First Sunday of Advent.
    // It will be always TEMPORA_PENT24_0, which will be placed either:
    // * instead of TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 23)
    // * directly after a week starting with TEMPORA_PENT23_0
    //   if the number of TEMPORA_PENT*_0 Sundays in given year == 24)
    // * directly after a week starting with TEMPORA_EPI6_0 (moved from post-epiphania period)
    //   if the number of TEMPORA_PENT*_0 Sundays in given year > 24)
    // """
    return previousSunday(new UTCDate(this.calcFirstAdventSunday(year)));
  }

  private calcSaturdayBefore24SundayAfterPentecost(year: number): Date {
    // """ Last Saturday before 24th Sunday after Pentecost.
    //
    // This is the end of potentially "empty" period that might appear
    // between 23rd and 24th Sunday after Pentecost if Easter is early.
    // In such case one or more Sundays after Epiphany (TEMPORA_EPI*_0) are moved here to "fill the gap"

    return previousSaturday(new UTCDate(this.calc24SundayAfterPentecost(year)));
  }

  private calcEmberWednesdaySeptember(year: number): Date {
    // """ Wednesday of the Ember Days of September.
    //
    // Ember Wednesday in September is a Wednesday after third Sunday
    // of September according to John XXIII's motu proprio
    // "Rubricarum instructum" of June 25 1960.
    // """
    let d = new UTCDate(year, 8, 1);
    while (getMonth(d) === 8) {
      if (isSunday(d) && getDate(d) >= 15 && getDate(d) <= 21) {
        break;
      }
      d = addDays(d, 1);
    }
    return nextWednesday(d);
  }

  private calcHolyName(year: number) {
    // """ The Feast of the Holy Name of Jesus.
    //
    // Kept on the First Sunday of the year; but if this Sunday falls on
    // 1st, 6th or 7th January, the feast is kept on 2nd January.
    // """
    let d = new UTCDate(year, 0, 1);

    while (getDate(d) <= 7) {
      if (isSunday(d)) {
        if (getDate(d) === 1 || getDate(d) === 6 || getDate(d) === 7) {
          return new UTCDate(year, 0, 2);
        }

        return d;
      }

      d = addDays(d, 1);
    }

    console.error("Unexpected condition: Holy Name date not found.");
  }

  private calcChristKing(year: number): Date {
    // The Feast of Christ the King, last Sunday of October.
    const d = new UTCDate(year, 10, 1);

    return previousSunday(d);
  }

  private calcSundayChristmasOctave(year: number): Date | null {
    // """
    // Sunday within the Octave of Christmas, falls between Dec 26 and Dec 31
    // """
    let d = new UTCDate(year, 11, 27); // December 27
    while (d.getFullYear() === year) {
      if (isSunday(d)) {
        return d;
      }
      d = addDays(d, 1);
    }
    return null;
  }

  get(date: string) {
    return this.container.get(date);
  }

  findDay(observanceId: string) {
    for (const [date_, day] of this.container) {
      if (day.all.some((observance) => observance.id === observanceId)) {
        return [date_, day];
      }
    }
  }

  private items() {
    return this.container.entries();
  }

  serialize(): { [date: string]: Day } {
    const serialized = {};
    for (const [date_, day] of this.items()) {
      serialized[date_] = day.serialize();
    }
    return serialized;
  }
}

export { Calendar };
