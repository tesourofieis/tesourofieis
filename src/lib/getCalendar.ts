import { Calendar } from "./src/calendar.ts";

function getCalendar(year: number) {
  const calendar = new Calendar(year);

  return calendar;
}

function getCalendarDay(year: number, date: string) {
  const calendar = new Calendar(year).get(date);

  return calendar;
}

export { getCalendar, getCalendarDay };
