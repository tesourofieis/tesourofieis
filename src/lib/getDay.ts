import { getYear } from "date-fns";
import { Calendar } from "./calendar.ts";
import { formatPropers } from "./utils.ts";

function getDay(date: string) {
  const calendar = new Calendar(getYear(date));

  const day = calendar.get(date);

  const proper = day?.getProper(calendar);

  const format = formatPropers(proper, day);

  return format;
}

export { getDay };
