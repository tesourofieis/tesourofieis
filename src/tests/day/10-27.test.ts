import { expect, test } from "vitest";
import { getDay } from "../../lib/getDay";

test("2024-10-27", () => {
  const proper = getDay("2024-10-27");

  expect(proper).toMatchSnapshot();
});
