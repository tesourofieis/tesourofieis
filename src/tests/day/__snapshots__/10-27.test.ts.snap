// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2024-10-27 1`] = `
[
  {
    "info": {
      "colors": [
        "w",
      ],
      "date": "2024-10-27",
      "day": Day {
        "celebration": [
          Observance {
            "colors": [
              "w",
            ],
            "date": "2024-10-27",
            "flexibility": "sancti",
            "id": "sancti:10-DUr:1:w",
            "name": "10-DUr",
            "priority": 13,
            "rank": 1,
            "title": "Cristo Rei",
            "weekday": 0,
          },
        ],
        "commemoration": [],
        "date": "2024-10-27",
        "tempora": [
          Observance {
            "colors": [
              "g",
            ],
            "date": "2024-10-27",
            "flexibility": "tempora",
            "id": "tempora:Pent23-0:2:g",
            "name": "Pent23-0",
            "priority": 45,
            "rank": 2,
            "title": "23º Domingo depois de Pentecostes",
            "weekday": 0,
          },
        ],
      },
      "id": "sancti:10-DUr:1:w",
      "rank": 1,
      "tempora": "23º Domingo depois de Pentecostes",
      "title": "Cristo Rei",
    },
    "sections": [
      {
        "body": {
          "latin": [
            "Apoc 5:12; 1:6",
            "Dignus est Agnus, qui occísus est, accípere virtútem, et divinitátem, et sapiéntiam, et fortitúdinem, et honórem. Ipsi glória et impérium in sǽcula sæculórum.",
            "Ps 71:1",
            "Deus, judícium tuum Regi da: et justítiam tuam Fílio Regis.",
            "Glória Patri…",
            "Dignus est Agnus, qui occísus est, accípere virtútem, et divinitátem, et sapiéntiam, et fortitúdinem, et honórem. Ipsi glória et impérium in sǽcula sæculórum.",
            "",
          ],
          "vernacular": [
            "Apoc 5:12; 1:6",
            "O Cordeiro, que foi imolado, é digno de receber o poder, a divindade, a sabedoria, a fortaleza, a honra: A Ele a glória e o império em todos os séculos dos séculos.",
            "Sl 71:1",
            "Ó Deus, dai ao Rei o poder de julgar; e a vossa justiça ao Filho do Rei.",
            "Glória ao Pai.",
            "O Cordeiro, que foi imolado, é digno de receber o poder, a divindade, a sabedoria, a fortaleza, a honra: A Ele a glória e o império em todos os séculos dos séculos.",
            "",
          ],
        },
        "id": "Introitus",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Omnípotens sempitérne Deus, qui in dilécto Fílio tuo, universórum Rege, ómnia instauráre voluísti: concéde propítius; ut cunctæ famíliæ géntium, peccáti vúlnere disgregátæ, ejus suavissímo subdántur império:",
            "Qui tecum…",
            "",
          ],
          "vernacular": [
            "Omnipotente e eterno Deus, que tudo quisestes restaurar no vosso dilecto Filho, Rei de todas as cousas, concedei-nos propício que todas as famílias do mundo, livres da chaga do pecado, se submetam ao Seu suavíssimo império.",
            "Que convosco vive...",
            "",
          ],
        },
        "id": "Oratio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Léctio Epístolæ beáti Pauli Apóstoli ad Colossénses.",
            "Col 1:12-20",
            "Fratres: Grátias ágimus Deo Patri, qui dignos nos fecit in partem sortis sanctórum in lúmine: qui erípuit nos de potestáte tenebrárum, et tránstulit in regnum Fílii dilectiónis suæ, in quo habémus redemptiónem per sánguinem ejus, remissiónem peccatórum: qui est imágo Dei invisíbilis, primogénitus omnis creatúra: quóniam in ipso cóndita sunt univérsa in cœlis et in terra, visibília et invisibília, sive Throni, sive Dominatiónes, sive Principátus, sive Potestátes: ómnia per ipsum, et in ipso creáta sunt: et ipse est ante omnes, et ómnia in ipso constant. Et ipse est caput córporis Ecclésiæ, qui est princípium, primogénitus ex mórtuis: ut sit in ómnibus ipse primátum tenens; quia in ipso complácuit omnem plenitúdinem inhabitáre; et per eum reconciliáre ómnia in ipsum, pacíficans per sánguinem crucis ejus, sive quæ in terris, sive quæ in cœlis sunt, in Christo Jesu Dómino nostro.",
            "",
          ],
          "vernacular": [
            "Lição da Ep.ª do B. Ap.º Paulo aos Colossenses.",
            "Col 1:12-20",
            "Meus irmãos: Damos graças a Deus Pai, porque nos fez dignos de participar da herança dos Santos na luz, nos livrou do poder das trevas e nos conduziu para o reino do seu muito amado Filho, no qual possuímos a redenção pelo seu sangue e a remissão dos pecados. Ele é a imagem de Deus invisível e o primogénito de todas as criaturas; porquanto n’Ele foram criadas todas as cousas nos céus e na terra, visíveis e invisíveis, quer sejam os Tronos, quer as Dominações, quer os Principados, quer as Potestades. Tudo foi criado por Ele e n’Ele próprio. Ele existe antes de todas as cousas e todas as cousas subsistem por Ele. Ele é a cabeça do corpo da Igreja e o princípio e o primogénito de todos os mortais, para que assim conserve a primazia de todas as cousas, pois foi do agrado do Pai que n’Ele residisse toda a plenitude e por Ele se reconciliem em si próprio todas as cousas, pacificando pelo seu sangue na Cruz tanto o que está na terra, como o que está no céu, em nosso Senhor Jesus Cristo.",
            "",
          ],
        },
        "id": "Lectio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 71:8; 78:11",
            "Dominábitur a mari usque ad mare, et a flúmine usque ad términos orbis terrárum.",
            "℣. Et adorábunt eum omnes reges terræ: omnes gentes sérvient ei. Allelúja, allelúja.",
            "Dan 7:14.",
            "Potéstas ejus, potéstas ætérna, quæ non auferétur: et regnum ejus, quod non corrumpétur. Allelúja.",
            "",
          ],
          "vernacular": [
            "Sl 71:8; 78:11",
            "Dominará de mar a mar e desde o rio aos confins da redondeza da terra! ℣. Adorá-l’O-ão todos os reis da terra: e todos os povos O servirão! Aleluia, aleluia.",
            "Dn 7:14",
            "Seu poder é eterno, nunca Lhe será tirado; seu reino é tal que nunca será corrompido. Aleluia.",
            "",
          ],
        },
        "id": "Graduale",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Sequéntia ☩ sancti Evangélii secúndum Joánnem.",
            "Joann 18:33-37",
            "In illo témpore: Dixit Pilátus ad Jesum: Tu es Rex Judæórum? Respóndit Jesus: A temetípso hoc dicis, an álii dixérunt tibi de me? Respóndit Pilátus: Numquid ego Judǽus sum? Gens tua et pontífices tradidérunt te mihi: quid fecísti? Respóndit Jesus: Regnum meum non est de hoc mundo. Si ex hoc mundo esset regnum meum, minístri mei útique decertárent, ut non tráderer Judǽis: nunc autem regnum meum non est hinc. Dixit ítaque ei Pilátus: Ergo Rex es tu? Respóndit Jesus: Tu dicis, quia Rex sum ego. Ego in hoc natus sum et ad hoc veni in mundum, ut testimónium perhíbeam veritáti: omnis, qui est ex veritáte, audit vocem meam.",
            "",
          ],
          "vernacular": [
            "Continuação ☩ do santo Evangelho segundo S. João.",
            "Jo 18:33-37",
            "Naquele tempo, disse Pilatos a Jesus: «Tu és o rei dos Judeus?». Respondeu Jesus: «Tu dizes isso de ti mesmo, ou foram outros que to disseram de mim?». Respondeu Pilatos: «Sou, porventura, judeu? Foram os teus compatriotas e os pontífices que te entregaram nas minhas mãos. O que fizeste?». Jesus respondeu: «Meu reino não é deste mundo. Se o meu reino fosse deste mundo, certamente os meus ministros haviam de pelejar para que Eu não fosse entregue aos judeus; porém, Eu o declaro agora, o meu reino não é daqui». Disse-Lhe, então, Pilatos: «És, portanto, rei?». Jesus respondeu: «Tu dizes que Eu sou rei. Eu para isso nasci e para isso vim a este mundo, a fim de dar testemunho da verdade. Todo aquele que é da verdade, escuta a minha voz».",
            "",
          ],
        },
        "id": "Evangelium",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 2:8.",
            "Póstula a me, et dabo tibi gentes hereditátem tuam, et possessiónem tuam términos terræ.",
            "",
          ],
          "vernacular": [
            "Sl 2:8",
            "Pede-me: e dar-te-ei as nações como herança; e como domínio os confins da terra!",
            "",
          ],
        },
        "id": "Offertorium",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Hóstiam tibi, Dómine, humánæ reconciliatiónis offérimus: præsta, quǽsumus; ut, quem sacrifíciis præséntibus immolámus, ipse cunctis géntibus unitátis et pacis dona concédat, Jesus Christus Fílius tuus, Dóminus noster:",
            "Qui tecum…",
            "",
          ],
          "vernacular": [
            "Vos oferecemos, Senhor, a hóstia da reconciliação humana, implorando-Vos que este mesmo Jesus Cristo, vosso Filho e Senhor nosso, o qual imolamos no presente sacrifício, conceda a todos os povos os benefícios da união e da paz:",
            "Que convosco vive...",
            "",
          ],
        },
        "id": "Secreta",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "de D.N. Jesu Christi Rege",
            "Vere dignum et justum est, æquum et salutáre, nos tibi semper et ubíque grátias ágere: Dómine sancte, Pater omnípotens, ætérne Deus: Qui unigénitum Fílium tuum, Dóminum nostrum Jesum Christum, Sacerdótem ætérnum et universórum Regem, óleo exsultatiónis unxísti: ut, seípsum in ara crucis hóstiam immaculátam et pacíficam ófferens, redemptiónis humánæ sacraménta perágeret: et suo subjéctis império ómnibus creatúris, ætérnum et universále regnum, imménsæ tuæ tráderet Majestáti. Regnum veritátis et vitæ: regnum sanctitátis et grátiæ: regnum justítiæ, amóris et pacis. Et ídeo cum Angelis et Archángelis, cum Thronis et Dominatiónibus cumque omni milítia cœléstis exércitus hymnum glóriæ tuæ cánimus, sine fine dicéntes:",
            "",
          ],
          "vernacular": [
            "de N. S. Jesus Cristo-Rei",
            "É verdadeiramente digno e justo, racional e salutar dar-Vos graças em todos os lugares e sempre, Senhor santo, Pai omnipotente, eterno Deus, que ungistes com o óleo da alegria o vosso Filho Unigénito, nosso Senhor Jesus Cristo, como Sacerdote eterno e Rei de todas as cousas, a fim de que, oferecendo-se na ara da Cruz - qual hóstia imaculada e pacífica -, realizasse as maravilhas da redenção humana e, ficando todas as criaturas sujeitas aoseu império, desse à vossa imensa majestade um reino eterno e universal: um reino de verdade e de vida; um reino de santidade e de graça; um reino de justiça, de amor e de paz. E por isso, com os Anjos e Arcanjos, com os Tronos e Dominações e com toda a milícia do exército celestial, cantamos um hino à vossa glória, dizendo incessantemente:",
            "",
          ],
        },
        "id": "Prefatio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Ps 28:10; 28:11",
            "Sedébit Dóminus Rex in ætérnum: Dóminus benedícet pópulo suo in pace.",
            "",
          ],
          "vernacular": [
            "Sl 28:10; 28:11",
            "Assentar-se-á, eternamente, o Senhor como Rei e abençoará o seu povo em paz.",
            "",
          ],
        },
        "id": "Communio",
        "subSections": [],
      },
      {
        "body": {
          "latin": [
            "Immortalitátis alimóniam consecúti, quǽsumus, Dómine: ut, qui sub Christi Regis vexíllis militáre gloriámur, cum ipso, in cœlésti sede, júgiter regnáre possímus:",
            "Qui tecum…",
            "",
          ],
          "vernacular": [
            "Tendo nós alcançado o alimento da imortalidade, Vos suplicamos, Senhor, permiti que possamos reinar perpetuamente na celestial mansão, juntamente com Jesus Cristo, sob cujas bandeiras nos gloriamos de militar:",
            "Que convosco vive...",
            "",
            "",
          ],
        },
        "id": "Postcommunio",
        "subSections": [],
      },
    ],
  },
]
`;
